### PR TITLE
Docker cross test infrastructure

### PR DESCRIPTION
**Pull request by posborne**
These commits contain the first portion of new infrastructure that will allow us to build for and test on a much larger set of platforms.  Nix and the tests are *not* compiling on many platforms right now, but I believe at this point that most of those problems are issues with nix (or libc) rather than the test infrastructure.

I want to get this out there so we can start filing issues and other people can start doing testing and development to resolve the issues popping up on various platforms.  None of this is integrated in with travis at this point -- that will be phase 3.  Phase 2 will be getting qemu going to run the tests for non-host platforms (this might be tricky when running in a container as `/dev/kvm` might need to be passed through).

This is based on the testing infrastructure in libc but modified quite a bit -- The libc infrastructure doesn't seem to be a great experience for testing locally.  If this approach works out, we might want to push this approach toward libc.

The Dockerfile/Image that goes along with this is currently here:
* https://github.com/posborne/docker-rust-cross
* https://hub.docker.com/r/posborne/rust-cross/

I would have no problem moving this into the org if that makes sense down the line.

### CODE REVIEW

**Comment by kamalmarhubi** [comment]
I am really excited to look at this!

**Comment by kamalmarhubi** [comment]
> (this might be tricky when running in a container as /dev/kvm might need to be passed through)

Could we prefer software emulation? That's what libc does for freebsd / netbsd / android on ARM. Travis can't even give us /dev/kvm if they wanted to: it's not available on GCE where they run. Or on EC2 for that matter...

File: ci/run.sh [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+echo "======================================================="
+echo "TESTING VERSION: ${VERSION}, TARGET: ${TARGET}"
+echo "======================================================="
+
+#
+# Tell cargo what linker to use and whatever else is required
+#
+configure_cargo() {
+  rm -rf .cargo
```
Comment (kamalmarhubi) [inline ci/run.sh]: this seems potentially distructive

File: ci/run.sh [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+# This is a hack as we cannot currently
+# ask cargo what test files it generated:
+# https://github.com/rust-lang/cargo/issues/1924
+find_binaries() {
+  target_base_dir="${BUILD_DIR}/${TARGET}/debug"
+
+  # find [[test]] sections and print the first line and
+  # hack it to what we want from there.  Also "nix" for
+  # tests that are implicitly prsent
```
Comment (kamalmarhubi) [inline ci/run.sh]: nit: present

File: ci/run.sh [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  for test_base in $( awk '/\[\[test\]\]/{getline; print}' "${MANIFEST_PATH}" | \
```
Comment (kamalmarhubi) [inline ci/run.sh]: you might be able to use `cargo read-manifest` and JSON processing here instead of fake-parsing the manifest

**Comment by kamalmarhubi** [comment]
Couple of comments so far. I am at a locale with slow internet, and so I can't play with it because pulling the image will take a while.

File: ci/run.sh (hunk 1 above)
Comment (posborne) [inline ci/run.sh]: Yes.   Currently, this is only executed in the container (and not in a directory that is part of a shared volume) -- May need to re-evaluate depending on how things evolve.

**Comment by posborne** [comment]
>> (this might be tricky when running in a container as /dev/kvm might need to be passed through)
> 
> Could we prefer software emulation? That's what libc does for freebsd / netbsd / android on ARM. Travis can't even give us /dev/kvm if they wanted to: it's not available on GCE where they run. Or on EC2 for that matter...

I think we can.  I am just getting around to actually testing the qemu bits and will give it a good try.  I'l have some time free next week (changing day jobs and have a week off), so I should have some time to round out testing and get things ready to go for travis.

File: ci/run.sh (hunk 1 above)
Comment (kamalmarhubi) [inline ci/run.sh]: I guess I mean that if I accidentally run this from the top level, and I have some cargo config, I will lose it. since I keep those out of git, I would then be sad.

Comment (kamalmarhubi) [inline ci/run.sh]: eg, I'd rather save the existing one, and have a trap to put the original back on exit.

**Comment by kamalmarhubi** [comment]
>  I am just getting around to actually testing the qemu bits and will give it a good try.

:+1: 

@posborne do you want to get this merged sooner than later?

**Comment by posborne** [comment]
@kamalmarhubi I am going to take one more pass at it, but then I would like to get something into master so people can start testing/fixing on various issues.  Phase 2 will be figuring out how best to integrate it with CI -- We will probably need to dump travis-rust-matrix, but that has been showing some strain lately anyhow.

I'm rebuilding the docker image right now as several dependencies were missing previously (the image is quite large now as the qemu-* packages are not small).

**Comment by posborne** [comment]
Ok, things seem to be working, so I think this is ready for a review.  I have the extra platforms tested set to not fail things right now until we get things sorted.  As we fix things, we can remove those versions from the `allow_failures` set.

r? @kamalmarhubi @fiveop

File: Cargo.toml [hunk 4]
```diff
@@ -20,7 +20,7 @@ preadv_pwritev = []
 signalfd = []
 
 [dependencies]
-libc     = "0.2.8"
+libc     = { git = "https://github.com/rust-lang/libc.git" }
 bitflags = "0.4"
```
Comment (fiveop) [inline Cargo.toml]: Why this change in this PR?

File: .travis.yml [hunk 5]
```diff
@@ -1,37 +1,57 @@
...
   - linux
   - osx
 
-env:
-  - RUST_TEST_THREADS=1 ARCH=x86_64
-  - RUST_TEST_THREADS=1 ARCH=i686
-
-script:
-  - curl -sSL https://raw.githubusercontent.com/carllerche/travis-rust-matrix/master/test | RUST_TEST_THREADS=1 bash
-  - cargo doc --no-deps
```
Comment (fiveop) [inline .travis.yml]: Did this deploy the docs somewhere? Or just build them?

File: Cargo.toml (hunk 4 above)
Comment (posborne) [inline Cargo.toml]: Ooops, I will remove this.  I had this locally for testing and it slipped in.

File: .travis.yml (hunk 5 above)
Comment (posborne) [inline .travis.yml]: Deployment is done later on and should be unchanged.

File: .travis.yml [hunk 6]
```diff
@@ -1,37 +1,57 @@
----
-sudo: false
 language: rust
+sudo: required
```
Comment (kamalmarhubi) [inline .travis.yml]: I really wish they'd let this be set per matrix entry... we're going to be slower to schedule because of this.

/me is planning to spend time on https://github.com/kamalmarhubi/containy-thing this week because of this silliness

File: ci/cargo-config [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+linker = "gcc"
+
+[target.x86_64-unknown-linux-musl]
+linker = "gcc"
+
+[target.i686-unknown-linux-gnu]
+linker = "gcc"
+
+[target.i686-unknown-linux-musl]
+linker = "gcc"
```
Comment (kamalmarhubi) [inline ci/cargo-config]: is this the default? if so we should maybe leave it out in case Rust ever changes to lld?

File: ci/README.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+determined by looking at the contents of the script.  The docker image
+used is [posborne/rust-cross][posborne/rust-cross].
+
+Running Test for Specific Architectures/Versions
+------------------------------------------------
+
+Suppose we have a failing test with Rust 1.6/1.7 on the raspberry pi.  In
+that case, we can run the following:
+
+    $ RUST_VERSIONS="1.6.0 1.7.0" RUST_TARGETS="arm-unknown-linux-gnueabihf" ci/run-all.sh
```
Comment (kamalmarhubi) [inline ci/README.md]: does this actually work? I don't see either variable referred to later

File: ci/run-travis.sh [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+# - TARGET: e.g. arm-unknown-linux-gnueabihf
+
+set -ex
+
+BASE_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
+
+if [ "$TRAVIS_OS_NAME" = "linux" ]; then
+  OS=unknown-linux-gnu
+else
+  OS=apple-darwin
```
Comment (kamalmarhubi) [inline ci/run-travis.sh]: can you maek this an `elif`, or the whole thing a `case` and abort if it's neither linux nor darwin?

File: ci/run.sh [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+BUILD_DIR="."
+
+VERSION="$1"
+TARGET="$2"
+
+export RUST_TEST_THREADS=1
+
+echo "======================================================="
+echo "TESTING VERSION: ${VERSION}, TARGET: ${TARGET}"
+echo "======================================================="
```
Comment (kamalmarhubi) [inline ci/run.sh]: minor: can you move this below function definitions?

File: ci/run.sh (hunk 3 above)
Comment (kamalmarhubi) [inline ci/run.sh]: if you get jq installed, then this works:

```
$ for bin in $(cargo read-manifest | jq '.targets[].name'); do echo $bin; done
"rustfmt"
"cargo-fmt"
"rustfmt"
"system"
```

Comment (kamalmarhubi) [inline ci/run.sh]: Ahhh I didn't actually read the linked bug. That is silly an annoying.

File: .travis.yml (hunk 6 above)
Comment (posborne) [inline .travis.yml]: Agreed.  It is annoying -- I removed some combos from the matrix to make overall build time somewhat reasonable, but that is a tradeoff and we could, for instance, miss a failure on some intermediate rust version.

File: ci/README.md (hunk 8 above)
Comment (posborne) [inline ci/README.md]: No, I'll need to update this doc.  There is an equivalent, but it would be something like this:

```
DOCKER_IMAGE=posborne/rust-cross:arm RUST_VERSION="1.7.0" RUST_TARGET="arm-unknown-linux-gnueabihf" ci/run-docker.sh
```

Right now the containers only container 1.7.0.  Adding other versions is not hard but it also starts to blow up the image size.  I figured stable was the most valuable version to test against if you are going to test against anything.

File: ci/run-travis.sh (hunk 9 above)
Comment (posborne) [inline ci/run-travis.sh]: Sure.  This bit was taken from libc, but I agree that it is probably better to be a bit more strict: https://github.com/rust-lang/libc/blob/master/ci/run-travis.sh#L11

File: ci/run.sh (hunk 3 above)
Comment (kamalmarhubi) [inline ci/run.sh]: It may still be possible to collect them up by doing `$bin{-,}*` or something like that

File: .travis.yml (hunk 6 above)
Comment (kamalmarhubi) [inline .travis.yml]: It was more a comment than anything else. It's been on my mind.

**Comment by kamalmarhubi** [comment]
awesome!!!

@homu r+
